File: cli/npm/goleo/runtime/battery/battery.go
//go:build !(android || ios) || goleo_battery

package battery

import "sync"

// BatteryInfo mirrors the web Battery Status API: Level is 0..1, and the
// time fields are seconds with -1 meaning unknown.
type BatteryInfo struct {
	Level           float64 `json:"level"`
	Charging        bool    `json:"charging"`
	ChargingTime    float64 `json:"chargingTime"`
	DischargingTime float64 `json:"dischargingTime"`
}

// Provider is a native battery backend. On mobile the shell registers one
// via SetProvider; on desktop the built-in platform implementation is used
// when no provider is set.
type Provider interface {
	GetBatteryInfo() (*BatteryInfo, error)
}

var (
	providerMu sync.RWMutex
	provider   Provider
)

func SetProvider(p Provider) {
	providerMu.Lock()
	defer providerMu.Unlock()
	provider = p
}

func GetBatteryInfo() (*BatteryInfo, error) {
	providerMu.RLock()
	p := provider
	providerMu.RUnlock()
	if p != nil {
		return p.GetBatteryInfo()
	}
	return platformGetBatteryInfo()
}

File: bridge/src/battery.ts
import { getBridge } from './bridge'

const bridge = () => getBridge()

export interface BatteryInfo {
  level: number
  charging: boolean
  chargingTime?: number
  dischargingTime?: number
}

export async function getBatteryInfo(): Promise<BatteryInfo> {
  try {
    return await bridge().invoke<BatteryInfo>('goleo:batteryGetInfo')
  } catch {
    if (typeof navigator !== 'undefined' && 'getBattery' in navigator) {
      const b = await (navigator as any).getBattery()
      return {
        level: b.level,
        charging: b.charging,
        chargingTime: b.chargingTime,
        dischargingTime: b.dischargingTime,
      }
    }
    throw new Error('battery API not available')
  }
}
